rust/src/vector.rs

```diff
@@ -0,0 +1,398 @@
+    /// vec.restore_array(array);
+    /// ```
+    pub fn get_array(&self, mtype: crate::MemType) -> ndarray::ArrayViewMut1<f64> {
+        let n = self.len();
+        let mut ptr = std::ptr::null_mut();
+        unsafe {
+            bind_ceed::CeedVectorGetArray(self.ptr, mtype as bind_ceed::CeedMemType, &mut ptr)
+        };
+        let ptr_slice: &mut [f64] = unsafe { slice::from_raw_parts_mut(ptr, n) };
+        ndarray::aview_mut1(ptr_slice)
+    }
+
+    /// Restore read/write access to a CeedVector via the specified memory type
+    ///
+    /// # arguments
+    ///
+    /// * 'array' - Array of vector data
+    ///
+    /// ```
+    /// # let ceed = ceed::Ceed::default_init();
+    /// let mut vec = ceed.vector(10);
+    /// let array = vec.get_array(ceed::MemType::Host);
+    /// vec.restore_array(array);
+    /// ```
+    pub fn restore_array(&self, array: ndarray::ArrayViewMut1<f64>) {
+        let mut ptr = std::ptr::null_mut();
+        unsafe { bind_ceed::CeedVectorRestoreArray(self.ptr, &mut ptr) };
+        drop(array);
+    }
+
+    /// Get read access to a CeedVector via the specified memory type
+    ///
+    /// # arguments
+    ///
+    /// * 'mtype' - Memory type on which to access the array
+    ///
+    /// ```
+    /// # let ceed = ceed::Ceed::default_init();
+    /// let mut vec = ceed.vector(10);
+    /// vec.set_value(42.0);
+    /// let array = vec.get_array_read(ceed::MemType::Host);
+    /// assert_eq!(array[5], 42.0);
+    /// vec.restore_array_read(array);
+    /// ```
+    pub fn get_array_read(&self, mtype: crate::MemType) -> ndarray::ArrayView1<f64> {
+        let n = self.len();
+        let mut ptr = std::ptr::null();
+        unsafe {
+            bind_ceed::CeedVectorGetArrayRead(self.ptr, mtype as bind_ceed::CeedMemType, &mut ptr)
+        };
+        let ptr_slice: &[f64] = unsafe { slice::from_raw_parts(ptr, n) };
+        ndarray::aview1(ptr_slice)
+    }
+
+    /// Restore read access to a CeedVector via the specified memory type
+    ///
+    /// # arguments
+    ///
+    /// * 'array' - Array of vector data
+    ///
+    /// ```
+    /// # let ceed = ceed::Ceed::default_init();
+    /// let vec = ceed.vector(10);
+    /// let array = vec.get_array_read(ceed::MemType::Host);
+    /// vec.restore_array_read(array);
+    /// ```
+    pub fn restore_array_read(&self, array: ndarray::ArrayView1<f64>) {
+        let mut ptr = std::ptr::null();
+        unsafe { bind_ceed::CeedVectorRestoreArrayRead(self.ptr, &mut ptr) };
+        drop(array);
+    }
+
+    /// Create an immutable view
+    ///
+    /// ```
+    /// # let ceed= ceed::Ceed::default_init();
+    /// let vec = ceed.vector_from_slice(&[10., 11., 12., 13.]);
+    /// let v = vec.view();
+    /// assert_eq!(v[0..2], [10., 11.]);
+    /// // It is valid to have multiple immutable views
+    /// let w = vec.view();
+    /// assert_eq!(v[1..], w[1..]);
+    /// ```
+    pub fn view(&self) -> VectorView {
+        VectorView::new(self)
+    }
```

Comment:
I'd like to get rid of the functions like `get_array`, which leak memory if the caller does not remember to `restore_array`. I don't have strong opinions about whether `VectorView` coerces (via `Deref`) to a slice or ndarray. One possible point in favor of slices is minimizing dependencies and encouraging users to use `zip` patterns for manipulating vectors in-place versus the common allocating semantics in `ndarray`.

Reply:
As thoroughly discussed on Slack, I dropped  `get_array` and `restore_array` and added a mutable analogue of this interface.